src/Chapter1.hs

```diff
@@ -542,7 +545,12 @@ value after "=" where the condition is true.
 Casual reminder about adding top-level type signatures for all functions :)
 -}
 
-mid x y z = error "mid: not implemented!"
+mid :: Int -> Int -> Int -> Int
+mid x y z
+    | x < y && x > z || x < z && x > y = x
+    | y < x && y > z || y < z && y > x = y
+    | z < y && z > x || z < x && z > y = z
+    | otherwise = error "mid: no middle found!"
```

Comment:
As we mentioned, the compiler in Haskell is very attentive to the exhaustive pattern-matching. And here it would warn you that `Pattern matching is not exhaustive`, as the guards have quite complicated logic, and the compiler won't be able to prove that it covers all the cases.

Because of that, you will need to add another guard – `| otherwise = z`, to tell the compiler, that your pattern matching is exhaustive 🙂


So, it looks like you can get rid of redundant `error` case and make the implementation shorter by returning `z` on the `otherwise` case and collapse last two guards into a single one.

Reply:
Thank you! Something I noticed with using guards directly in a function is that `=` doesn't follow right after the function name and arguments? For example, the following doesn't compile due to the `=` in `mid x y z =`:
```haskell
mid :: Int -> Int -> Int -> Int
mid x y z =
    | x < y && x > z || x < z && x > y = x
    | y < x && y > z || y < z && y > x = y
    | otherwise = z
```
This was interesting to me since guards look similar to pattern matching in another language I'm familiar with, OCaml. I'm guessing it's because we're doing a pattern matching on the function arguments `x`, `y`, `z`? Is this syntax kind of like a shortcut for saying:
```
mid x y z =
    match (x, y, z)
        | ...
        | ...
        | ...
```
?

Reply:
@Alison-Li Yes, your intuition is correct. We are exactly doing pattern-matching on variables, but it's a special case of pattern matching — a pattern that captures anything and just assigns the variable name to it. The `=` sign goes after each guard because Haskell validates all checks from top to  bottom and returns the expression on a first guard that succeeds.